### ISSUE TITLE
Why do I get ChallengeChoice.EMAIL every time when I log in to get the SMS code?

### ISSUE THREAD

**Issue by tmxd09887** [subzeroid/instagrapi#1969]
got ChallengeChoice.EMAIL

But in fact, I received a login sms verification code

When I entered the SMS code, I got a login error message

And my account is not linked to Email

Is ChallengeChoice set to Email by default? How can I change it back to SMS?

**Comment by jbosolutions** [comment]
Me too… and Instagram ask to change password

**Comment by tmxd09887** [comment]
I guess this is the case. Instagram asked for an SMS verification code, but sent back an email verification code, thus triggering risk control (requiring a password change). The normal process should be to successfully log in after entering the SMS verification code.

**Comment by jbosolutions** [comment]
Yeah I disabled sms auth. But I can’t disable email auth code… also I don’t know why Instagram ask me to change password, no sense

**Comment by jbosolutions** [comment]
Do you know how disable all auto codes on Instagram? Did you resolved it?

**Comment by tmxd09887** [comment]
sms auth is required by Instagram, I don't think it can be disabled

**Comment by jbosolutions** [comment]
This is the error I get when I put the verification code in console:

ChallengeResolveMixin.challenge_resolve() takes 2 positional arguments but 3 were given


Maybe because I have another account using Instagrapi in same computer?

**Comment by tmxd09887** [comment]
I don't know how to disable it, but there must be something wrong with the verification code processing.

**Comment by sunmasters** [comment]
It means, you should save session data. Reuse it again to avoid asking OTP again

**Comment by tmxd09887** [comment]
> It means, you should save session data. Reuse it again to avoid asking OTP again

Hello, this is the first time logging in, there is no existing session data

**Comment by sunmasters** [comment]
Is device data is same?

**Comment by tmxd09887** [comment]
The device is generated by instagrapi, so it is definitely different from the one used when registering. But the key point is that there is a bug in the process of submitting the SMS verification code.

**Comment by sunmasters** [comment]
Maybe API need update. You have to update in your system

**Comment by tmxd09887** [comment]
> Maybe API need update. You have to update in your system .
> 
> I think you can give user agent as your device?

Yes, this could be a way.

**Comment by subzeroid** [comment]
`ChallengeChoice` is selected from the challenge step Instagram returns. It is not a setting that can reliably force SMS for every login.

For older challenge flows, instagrapi may fall back to `EMAIL` when Instagram does not expose a clear SMS-specific step. For newer login verification, many accounts are moved to the CAA/Bloks two-factor flow, where the available methods come from Instagram server state. If Instagram sends an SMS but the old challenge endpoint reports a different step, the safest path is to use the current 2FA/Bloks flow or verify the account manually in the official app.

Recommended handling:

- save and reuse session settings so the account is not challenged on every run
- keep device/proxy/country/locale stable for the account
- use `login(..., verification_code="123456")` for 2FA codes when possible
- use the Bloks two-factor helpers if Instagram exposes `two_step_verification_context`
- avoid switching identity while a challenge is in progress

Docs:

- Challenge resolver: https://subzeroid.github.io/instagrapi/usage-guide/challenge_resolver.html
- Bloks two-factor flow: https://subzeroid.github.io/instagrapi/usage-guide/totp.html#bloks-two-factor-flow
- Best practices: https://subzeroid.github.io/instagrapi/usage-guide/best-practices.html
